File: PianoKeyboard_java_eclipse/src/com/teuskim/pianokeyboard/RegisterCustomActivity.java
package com.teuskim.pianokeyboard;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.res.Resources;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.teuskim.pianokeyboard.PianoKeyboard.Key;
import com.teuskim.pianokeyboard.PianoKeyboardView.OnKeyboardActionListener;

public class RegisterCustomActivity extends BaseActivity implements OnClickListener,OnKeyboardActionListener {
	
	private EditText mNameCustom;
	private PianoKeyboardView mKeyboardView;
	private LinearLayout mKeymapListView;
	private Button mBtnRecommend;
	private Button mBtnReset;
	
	private Map<Integer, String> mKeymap;
	private int mPressedIndex;
	private PianoKeyboard mPianoKeyboard;
	private int mCustomId;
	private PianoKeyboardDb mDb;
	private LayoutInflater mInflater;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.register_custom);
		findViews();
		
		mDb = PianoKeyboardDb.getInstance(getApplicationContext());
		mKeymap = new HashMap<Integer, String>();
		mInflater = LayoutInflater.from(this);
		
		mCustomId = getIntent().getIntExtra("customId", 0);
		if(mCustomId > 0){
			String name = mDb.getCustomKeySetName(mCustomId);
			mNameCustom.setText(name);
			
			List<PianoKeyboardDb.CustomKeysetData> cksdList = mDb.getCustomKeySetDataList(mCustomId);
			for(PianoKeyboardDb.CustomKeysetData cksd : cksdList){
				mKeymap.put(cksd.mPosition, cksd.mData);
			}
			
			refreshKeymap();
		}
		
		// google analytics
		mTracker.trackPageView("RegisterCustomActivity");
	}

	protected void findViews(){
		mNameCustom = (EditText) findViewById(R.id.name_custom);
		mKeyboardView = (PianoKeyboardView) findViewById(R.id.keyboard_view);
		mKeyboardView.setIsRegisterMode(true);
		mKeyboardView.adjustLayoutParams(3,3,1);
		mKeyboardView.setOnKeyboardActionListener(this);
		mKeymapListView = (LinearLayout) findViewById(R.id.keymap_list);
		mBtnRecommend = (Button) findViewById(R.id.btn_custom_recommend);
		mBtnReset = (Button) findViewById(R.id.btn_custom_reset);
		
		mKeyboardView.setOnClickListener(this);
		findViewById(R.id.btn_save).setOnClickListener(this);
		findViewById(R.id.btn_cancel).setOnClickListener(this);
		mBtnRecommend.setOnClickListener(this);
		mBtnReset.setOnClickListener(this);
	}

	@Override
	public void onClick(View v) {
		switch(v.getId()){
		case R.id.btn_save:
			save();
			break;
		case R.id.btn_cancel:
			finish();
			break;
		case R.id.btn_custom_recommend:
			dialogRecommendedCustom();
			break;
		case R.id.btn_custom_reset:
			reset();
			break;
		}
	}
	
	private void save(){
		String name = mNameCustom.getText().toString();
		if(name == null || name.length() == 0){
			Toast.makeText(getApplicationContext(), R.string.toast_input_title, Toast.LENGTH_SHORT).show();
			return;
		}
		
		SaveTask task = new SaveTask();
		task.execute(name);
		
		finish();
	}
	
	private class SaveTask extends AsyncTask<String, Integer, Boolean>{

		@Override
		protected Boolean doInBackground(String... params) {
			String name = params[0];
			String showYN = "N";
			if(mCustomId > 0){
				showYN = mDb.getCustomKeySetShowYN(mCustomId);
				if(mDb.deleteCustomKeyset(mCustomId) == false){
					mDb.deleteCustomKeyset(mCustomId);
				}
			}		
			return mDb.insertCustomKeyset(name, showYN, mKeymap);
		}

		@Override
		protected void onPostExecute(Boolean result) {
			if(result)
				Toast.makeText(getApplicationContext(), R.string.toast_save_ok, Toast.LENGTH_SHORT).show();
			else
				Toast.makeText(getApplicationContext(), R.string.toast_save_fail, Toast.LENGTH_SHORT).show();
		}
		
	}

	@Override
	public void onTouchDown(int keyType, int index, Key key) {
		if(keyType == KeyView.KEY_TYPE_WHITE){
			mPressedIndex = index;
		}
		else{
			mPressedIndex = index + PianoKeyboard.WHITE_NUM;
		}
	}

	@Override
	public void onTouchMove() {
	}

	@Override
	public void onTouchUp() {
		dialogKeyInput();
	}
	
	private void dialogKeyInput(){
		LayoutInflater inflater = LayoutInflater.from(getApplicationContext());
		View layout = inflater.inflate(R.layout.register_custom_dialog, null);
		final EditText edittext = (EditText) layout.findViewById(R.id.input_keyset);
		
		new AlertDialog.Builder(this)
			.setTitle(R.string.title_input_key)
			.setView(layout)
			.setPositiveButton(R.string.btn_ok, new DialogInterface.OnClickListener() {
				
				public void onClick(DialogInterface dialog, int which) {
					String data = edittext.getText().toString();
					if(data != null && data.length() > 0){
						mKeymap.put(mPressedIndex, data);
						
						refreshKeymap();
					}
				}
			})
			.create().show();
	}
	
	private void refreshKeymap(){
		mKeymapListView.removeAllViews();
		
		TreeSet<Integer> keyset = new TreeSet<Integer>(mKeymap.keySet());
		Iterator<Integer> iter = keyset.iterator();
		boolean hasKeymap = false;
		boolean isShow1 = false;
		boolean isShow2 = false;
		boolean isShow3 = false;
		boolean isShow4 = false;
		
		while(iter.hasNext()){
			final int index = iter.next();
			
			if(index < 8 && isShow1 == false){
				addCustomSubtitle(R.string.txt_piano_white_top);
				isShow1 = true;
			}
			else if(index >= 8 && index < 16 && isShow2 == false){
				mInflater.inflate(R.layout.register_custom_divider, mKeymapListView);
				addCustomSubtitle(R.string.txt_piano_white_bottom);
				isShow2 = true;
			}
			else if(index >= 16 && index < 22 && isShow3 == false){
				mInflater.inflate(R.layout.register_custom_divider, mKeymapListView);
				addCustomSubtitle(R.string.txt_piano_black_top);
				isShow3 = true;
			}
			else if(index >= 22 && isShow4 == false){
				mInflater.inflate(R.layout.register_custom_divider, mKeymapListView);
				addCustomSubtitle(R.string.txt_piano_black_bottom);
				isShow4 = true;
			}
			
			View v = mInflater.inflate(R.layout.register_custom_item, null);
			TextView customData = (TextView) v.findViewById(R.id.custom_data);
			customData.setText(mKeymap.get(index));
			
			v.findViewById(R.id.btn_delete).setOnClickListener(new OnClickListener() {
				
				@Override
				public void onClick(View v) {
					mKeymap.remove(index);
					refreshKeymap();
				}
			});
			
			mKeymapListView.addView(v);
			hasKeymap = true;
		}
		
		if(hasKeymap)
			mBtnReset.setVisibility(View.VISIBLE);
		else
			mBtnReset.setVisibility(View.GONE);
		
		mPianoKeyboard = new PianoKeyboard(this, mKeymap);
		mKeyboardView.setKeyboard(mPianoKeyboard);
	}
	
	private void addCustomSubtitle(int resId){
		View titleView = mInflater.inflate(R.layout.register_custom_title_item, null);
		TextView titleTextView = (TextView) titleView.findViewById(R.id.custom_title);
		titleTextView.setText(resId);
		mKeymapListView.addView(titleView);
	}
	
	private void reset(){
		mNameCustom.setText(null);
		mKeymap.clear();
		refreshKeymap();
	}
	
	private void dialogRecommendedCustom(){
		new AlertDialog.Builder(this)
			.setTitle(R.string.btn_custom_recommend)
			.setItems(R.array.custom_names, new DialogInterface.OnClickListener() {
				
				@Override
				public void onClick(DialogInterface dialog, int which) {
					Resources res = getResources();
					String[] names = res.getStringArray(R.array.custom_names);
					mNameCustom.setText(names[which]);
					
					String[] strs = null;
					switch(which){
					case 0:
						strs = res.getStringArray(R.array.ex1_symbols);
						break;
					case 1:
						strs = res.getStringArray(R.array.ex2_emoticons);
						break;
					case 2:
						strs = res.getStringArray(R.array.ex3_pitch_names);
						break;
					case 3:
						strs = res.getStringArray(R.array.ex4_hangul_2nd);
						break;
					}
					if(strs != null){
						for(int i=0; i<strs.length; i++){
							mKeymap.put(i, strs[i]);
						}
					}
					refreshKeymap();
				}
			})
			.create().show();
	}
	
}

File: PianoKeyboard_kotlin/app/src/main/java/com/teuskim/pianokeyboard/RegisterCustomActivity.kt
package com.teuskim.pianokeyboard

import java.util.HashMap
import java.util.TreeSet

import android.app.AlertDialog
import android.os.AsyncTask
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast

import com.teuskim.pianokeyboard.PianoKeyboard.Key
import com.teuskim.pianokeyboard.PianoKeyboardView.OnKeyboardActionListener

class RegisterCustomActivity : BaseActivity(), OnClickListener, OnKeyboardActionListener {

    private var mNameCustom: EditText? = null
    private var mKeyboardView: PianoKeyboardView? = null
    private var mKeymapListView: LinearLayout? = null
    private var mBtnRecommend: Button? = null
    private var mBtnReset: Button? = null

    private var mKeymap: MutableMap<Int, String>? = null
    private var mPressedIndex: Int = 0
    private var mPianoKeyboard: PianoKeyboard? = null
    private var mCustomId: Int = 0
    private var mDb: PianoKeyboardDb? = null
    private var mInflater: LayoutInflater? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.register_custom)
        findViews()

        mDb = PianoKeyboardDb.getInstance(applicationContext)
        mKeymap = HashMap()
        mInflater = LayoutInflater.from(this)

        mCustomId = intent.getIntExtra("customId", 0)
        if (mCustomId > 0) {
            val name = mDb!!.getCustomKeySetName(mCustomId)
            mNameCustom!!.setText(name)

            val cksdList = mDb!!.getCustomKeySetDataList(mCustomId)
            for (cksd in cksdList) {
                mKeymap!!.put(cksd.mPosition, cksd.mData!!)
            }

            refreshKeymap()
        }

        // google analytics
        mTracker!!.trackPageView("RegisterCustomActivity")
    }

    protected fun findViews() {
        mNameCustom = findViewById<View>(R.id.name_custom) as EditText
        mKeyboardView = findViewById<View>(R.id.keyboard_view) as PianoKeyboardView
        mKeyboardView!!.setIsRegisterMode(true)
        mKeyboardView!!.adjustLayoutParams(3.0, 3.0, 1.0)
        mKeyboardView!!.setOnKeyboardActionListener(this)
        mKeymapListView = findViewById<View>(R.id.keymap_list) as LinearLayout
        mBtnRecommend = findViewById<View>(R.id.btn_custom_recommend) as Button
        mBtnReset = findViewById<View>(R.id.btn_custom_reset) as Button

        mKeyboardView!!.setOnClickListener(this)
        findViewById<View>(R.id.btn_save).setOnClickListener(this)
        findViewById<View>(R.id.btn_cancel).setOnClickListener(this)
        mBtnRecommend!!.setOnClickListener(this)
        mBtnReset!!.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_save -> save()
            R.id.btn_cancel -> finish()
            R.id.btn_custom_recommend -> dialogRecommendedCustom()
            R.id.btn_custom_reset -> reset()
        }
    }

    private fun save() {
        val name = mNameCustom!!.text.toString()
        if (name == null || name.length == 0) {
            Toast.makeText(applicationContext, R.string.toast_input_title, Toast.LENGTH_SHORT).show()
            return
        }

        val task = SaveTask()
        task.execute(name)

        finish()
    }

    private inner class SaveTask : AsyncTask<String, Int, Boolean>() {

        override fun doInBackground(vararg params: String): Boolean? {
            val name = params[0]
            var showYN = "N"
            if (mCustomId > 0) {
                showYN = mDb!!.getCustomKeySetShowYN(mCustomId)
                if (mDb!!.deleteCustomKeyset(mCustomId.toLong()) == false) {
                    mDb!!.deleteCustomKeyset(mCustomId.toLong())
                }
            }
            return mDb!!.insertCustomKeyset(name, showYN, mKeymap)
        }

        override fun onPostExecute(result: Boolean?) {
            if (result!!)
                Toast.makeText(applicationContext, R.string.toast_save_ok, Toast.LENGTH_SHORT).show()
            else
                Toast.makeText(applicationContext, R.string.toast_save_fail, Toast.LENGTH_SHORT).show()
        }

    }

    override fun onTouchDown(keyType: Int, index: Int, key: Key) {
        if (keyType == KeyView.KEY_TYPE_WHITE) {
            mPressedIndex = index
        } else {
            mPressedIndex = index + PianoKeyboard.WHITE_NUM
        }
    }

    override fun onTouchMove() {}

    override fun onTouchUp() {
        dialogKeyInput()
    }

    private fun dialogKeyInput() {
        val inflater = LayoutInflater.from(applicationContext)
        val layout = inflater.inflate(R.layout.register_custom_dialog, null)
        val edittext = layout.findViewById<View>(R.id.input_keyset) as EditText

        AlertDialog.Builder(this)
                .setTitle(R.string.title_input_key)
                .setView(layout)
                .setPositiveButton(R.string.btn_ok) { dialog, which ->
                    val data = edittext.text.toString()
                    if (data != null && data.length > 0) {
                        mKeymap!![mPressedIndex] = data

                        refreshKeymap()
                    }
                }
                .create().show()
    }

    private fun refreshKeymap() {
        mKeymapListView!!.removeAllViews()

        val keyset = TreeSet(mKeymap!!.keys)
        val iter = keyset.iterator()
        var hasKeymap = false
        var isShow1 = false
        var isShow2 = false
        var isShow3 = false
        var isShow4 = false

        while (iter.hasNext()) {
            val index = iter.next()

            if (index < 8 && isShow1 == false) {
                addCustomSubtitle(R.string.txt_piano_white_top)
                isShow1 = true
            } else if (index >= 8 && index < 16 && isShow2 == false) {
                mInflater!!.inflate(R.layout.register_custom_divider, mKeymapListView)
                addCustomSubtitle(R.string.txt_piano_white_bottom)
                isShow2 = true
            } else if (index >= 16 && index < 22 && isShow3 == false) {
                mInflater!!.inflate(R.layout.register_custom_divider, mKeymapListView)
                addCustomSubtitle(R.string.txt_piano_black_top)
                isShow3 = true
            } else if (index >= 22 && isShow4 == false) {
                mInflater!!.inflate(R.layout.register_custom_divider, mKeymapListView)
                addCustomSubtitle(R.string.txt_piano_black_bottom)
                isShow4 = true
            }

            val v = mInflater!!.inflate(R.layout.register_custom_item, null)
            val customData = v.findViewById<View>(R.id.custom_data) as TextView
            customData.text = mKeymap!![index]

            v.findViewById<View>(R.id.btn_delete).setOnClickListener {
                mKeymap!!.remove(index)
                refreshKeymap()
            }

            mKeymapListView!!.addView(v)
            hasKeymap = true
        }

        if (hasKeymap)
            mBtnReset!!.visibility = View.VISIBLE
        else
            mBtnReset!!.visibility = View.GONE

        mPianoKeyboard = PianoKeyboard(this, mKeymap!!)
        mKeyboardView!!.keyboard = mPianoKeyboard
    }

    private fun addCustomSubtitle(resId: Int) {
        val titleView = mInflater!!.inflate(R.layout.register_custom_title_item, null)
        val titleTextView = titleView.findViewById<View>(R.id.custom_title) as TextView
        titleTextView.setText(resId)
        mKeymapListView!!.addView(titleView)
    }

    private fun reset() {
        mNameCustom!!.text = null
        mKeymap!!.clear()
        refreshKeymap()
    }

    private fun dialogRecommendedCustom() {
        AlertDialog.Builder(this)
                .setTitle(R.string.btn_custom_recommend)
                .setItems(R.array.custom_names) { dialog, which ->
                    val res = resources
                    val names = res.getStringArray(R.array.custom_names)
                    mNameCustom!!.setText(names[which])

                    var strs: Array<String>? = null
                    when (which) {
                        0 -> strs = res.getStringArray(R.array.ex1_symbols)
                        1 -> strs = res.getStringArray(R.array.ex2_emoticons)
                        2 -> strs = res.getStringArray(R.array.ex3_pitch_names)
                        3 -> strs = res.getStringArray(R.array.ex4_hangul_2nd)
                    }
                    if (strs != null) {
                        for (i in strs.indices) {
                            mKeymap!![i] = strs[i]
                        }
                    }
                    refreshKeymap()
                }
                .create().show()
    }

}
